scripts/performance/benchmark_utils.py

```diff
@@ -0,0 +1,288 @@
+        path = os.path.join(result_dir, f)
+        if os.path.isfile(path):
+            summarize_args.append(path)
+
+    with open(os.path.join(summary_dir, 'summary.txt'), 'wb') as f:
+        subprocess.check_call(summarize_args, stdout=f)
+
+
+def _get_s3transfer_performance_script(script_name):
+    """Retrieves an s3transfer performance script if available."""
+    s3transfer_directory = os.path.dirname(s3transfer.__file__)
+    s3transfer_directory = os.path.dirname(s3transfer_directory)
+    scripts_directory = 'scripts/performance'
+    scripts_directory = os.path.join(s3transfer_directory, scripts_directory)
+    script = os.path.join(scripts_directory, script_name)
+
+    if os.path.isfile(script):
+        return script
+    else:
+        return None
+
+
+def get_benchmark_script():
+    return _get_s3transfer_performance_script('benchmark')
+
+
+def get_summarize_script():
+    return _get_s3transfer_performance_script('summarize')
+
+
+def backup(source, recursive):
+    """Backup a given source to a temporary location.
+
+    :type source: str
+    :param source: A local path or s3 path to backup.
+
+    :type recursive: bool
+    :param recursive: if True, the source will be treated as a directory.
+    """
+    if source[:5] == 's3://':
+        parts = source.split('/')
+        parts.insert(3, str(uuid.uuid4()))
+        backup_path = '/'.join(parts)
+    else:
+        name = os.path.split(source)[-1]
+        temp_dir = tempfile.mkdtemp()
+        backup_path = os.path.join(temp_dir, name)
+
+    copy(source, backup_path, recursive)
+    return backup_path
+
+
+def copy(source, destination, recursive):
+    """Copy files from one location to another.
+
+    The source and destination must both be s3 paths or both be local paths.
+
+    :type source: str
+    :param source: A local path or s3 path to backup.
+
+    :type destination: str
+    :param destination: A local path or s3 path to backup the source to.
+
+    :type recursive: bool
+    :param recursive: if True, the source will be treated as a directory.
+    """
+    if 's3://' in [source[:5], destination[:5]]:
+        cp_args = ['aws', 's3', 'cp', source, destination, '--quiet']
+        if recursive:
+            cp_args.append('--recursive')
+        subprocess.check_call(cp_args)
+        return
+
+    if recursive:
+        shutil.copytree(source, destination)
+    else:
+        shutil.copy(source, destination)
+
+
+def clean(destination, recursive):
+    """Delete a file or directory either locally or on S3."""
+    if destination[:5] == 's3://':
+        rm_args = ['aws', 's3', 'rm', '--quiet', destination]
+        if recursive:
+            rm_args.append('--recursive')
+        subprocess.check_call(rm_args)
+    else:
+        if recursive:
+            shutil.rmtree(destination)
+        else:
+            os.remove(destination)
+
+
+def get_random_subfolder(destination):
+    """Get a random subdirectory of a given directory."""
```

Comment:
Might be worth clarifying/renaming this to make it clear we will *create* a new random subfolder under the provided destination for local files (vs. randomly picking an existing subdirectory).